machina/apps/forum_permission/handler.py

```diff
@@ -329,144 +326,21 @@ def _get_forums_for_user(self, user, perm_codenames, use_tree_hierarchy=False):
-                per_forum_granted_group_perms = list(
-                    filter(lambda p: p.has_perm and p.forum_id is not None, group_perms)
-                )
-                # A permission can be granted on user-forum level, and that takes precedence over
-                # nongranted group permissions so we do not add those to the list.
-                per_forum_nongranted_group_perms = list(
-                    filter(lambda p: not p.has_perm and p.forum_id is not None and
-                           p.forum_id not in granted_user_forum_ids, group_perms)
-                )
-
-                # Now we can update the list of forums ids for which permissions are explicitly not
-                # granted.
-                granted_group_forum_ids = [p.forum_id for p in per_forum_granted_group_perms]
-                nongranted_forum_ids += [
-                    p.forum_id for p in per_forum_nongranted_group_perms
-                    if p.forum_id not in granted_group_forum_ids
-                ]
-
-                # Now we will update our previous dictionary that associated each forum ID with a
-                # set of granted permissions (at the user level). We will update it with the new
-                # permissions we've computed for the user's groups.
-                for perm in per_forum_granted_group_perms:
-                    granted_permissions_per_forum[perm.forum_id].add(perm.permission_id)
-                for forum_id in initial_forum_ids:
-                    granted_permissions_per_forum[forum_id].update(
-                        [perm.permission_id for perm in globally_granted_group_perms]
-                    )
-
-            # We keep only the forum IDs for which the length of the set containing the granted
-            # permissions is equal to the number of initial permission codenames. The other forums
-            # have not all the required granted permissions, so we just throw them away.
-            for forum_id in list(granted_permissions_per_forum):
-                if len(granted_permissions_per_forum[forum_id]) < required_perm_codenames_count:
-                    del granted_permissions_per_forum[forum_id]
-
-            # Alright! It is now possible to filter the initial queryset using the forums associated
-            # with the granted permissions and the list of forums for which permissions are
-            # explicitly not granted.
-            forum_objects = [
-                f for f in forums
-                if f.id in granted_permissions_per_forum and f.id not in nongranted_forum_ids
-            ]
-
-            if (
-                not user.is_anonymous and
-                set(perm_codenames).issubset(
-                    set(machina_settings.DEFAULT_AUTHENTICATED_USER_FORUM_PERMISSIONS)
-                )
-            ):
-                forum_objects += [
-                    f for f in forums if f.id not in nongranted_forum_ids and f not in forum_objects
-                ]
```

Comment:
Why remove all this logic? There is a reason for this specific implementation: it was here in order to retrieve _all_ forums that match a specific set of permissions for a specific user. On the other hand `ForumPermissionChecker` instances work on a per-forum basis.

As a rule of thumb, the `ForumPermissionHandler` class defines logics to filter lists of forums for a specific user based on permissions while the `ForumPermissionChecker` class defines logics to tell whether a specific forum has a specific set of permissions granted for a specific user.

That's why I'd prefer if the new `authenticated_user` rule could be implemented as part of the original implementation of this `_get_forums_for_user` method. Otherwise you're forced to use the checker on each forum instance (like you did), and this generates _a lot_ of SQL queries.

For example, on my test project, displaying the index results in about 10 SQL queries in the current _master_ branch. With your modifications I end up with **36** SQL queries to display the index for an authenticated user, which not acceptable.

So here is what I can suggest:

1. you introduce the new `authenticated_user` logic to the `_get_forums_for_user` by complying to the existing strategy
2. and you introduce the new `authenticated_user` logic to the `ForumPermissionChecker` too

Reply:
I removed this logic because I noticed that there was a lot of duplication in the code of checker and handler when I wanted to add the authenticated_user logic. But there were also some differences that made the outcome of a check different between both files while you'd expect the same result (it is all about the same permissions...). So to prevent more differentiation and duplicate code in the future I though it wise to have the permission checking code just in one place. 
However, I did not take into account the increase in SQL queries. I'll be looking into that

Reply:
I addressed this issue and committed new code that still has all the permission-collecting logic in one place but does no longer require a query per loop iteration. The amount of SQL queries for master and this branch are now the same on my test environment.

Reply:
> I addressed this issue and committed new code that still has all the permission-collecting logic in one place but does no longer require a query per loop iteration. The amount of SQL queries for master and this branch are now the same on my test environment.

I still have a difference between your branch and master.

In order to get this PR merged faster, I would suggest you to keep using the existing strategy (without touching so much at the `PermissionHandler` class, just add what's necessary in it to support the new authenticated user mechanism).

And in a second PR you could introduce the refactoring to the `PermissionHandler` / `ForumPermissionChecker`. Since this refactoring has some implications regarding performances and SQL queries I would prefer if we could split things that way.

What do you think?